Add AppendingByteSerializer class

oulgen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148227
* __->__ #148226

This PR adds a new util class that enables efficient appending of sequential byte data with custom serialization and deserialization.

oulgen committed:
Add AppendingByteSerializer class

[ghstack-poisoned]

oulgen committed:
None on "Add AppendingByteSerializer class"


This PR adds a new util class that enables efficient appending of sequential byte data with custom serialization and deserialization. 

[ghstack-poisoned]

oulgen commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+#######################################
+
+
+class BytesWriter:
```
What benefit would we get from using dataclass here? generally i use dataclasses when i want hash/eq/str etc for free but in this case i dont need any of that?

Skylion007 replied:
Don't need it now, but useful for debugging later. Also you can remove the need for a custom `__init__` function.

aorenste commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+        self._data = bytearray()
+
+    def write_int(self, i: int) -> None:
+        self._data.extend(i.to_bytes(8, byteorder="big", signed=False))
```
Why byteorder "big"? x86 native is "little". Also these values are the defaults - if you're calling them out for a reason then there should probably be a comment.

oulgen replied:
unfortunately py3.9 does not have the defaults.. so tests were failing

aorenste commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+        self.clear()
+
+    def clear(self) -> None:
+        # Use first byte as version
```
Comment says "byte" but code uses a uint64 (and since you're currently writing in big endian it's not even coincidentally the first byte - it's the 8th).

oulgen replied:
good catch, this was the fault of refactoring

aorenste commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+    def get(self) -> bytes:
+        return self._writer.get()
+
+    def appendListAndGet(self, elems: Iterable[T]) -> bytes:
```
nit: This feels like either this should be `append_list_and_get` or the next function should be `toList`...

Also in pythonic this would be named more like `extend`, not `appendList`.

I wouldn't combine the get here - I'd just have an `extend` and if someone wants to immediate get then let them use the `get` function as a second call.

aorenste commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+    def write_bytes(self, b: bytes) -> None:
+        self._data.extend(b)
+
+    def get(self) -> bytes:
```
to_bytes? (and if so then AppendingByteSerializer.get should match)

aorenste commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+        return self.get()
+
+    @staticmethod
+    def to_list(data: bytes, *, deserialize_fn: Callable[[bytes], T]) -> list[T]:
```
Maybe this should just be a top-level function? Not sure it makes sense as a static...

oulgen replied:
I wanted it to be associated with the class so that it is clear that it is meant only for this class?

aorenste reviewed (approved):
I wonder if instead of passing around bytes you should pass a BytesWriter to the serializer function and a BytesReader to the deserializer function and have it read/write bytes straight into the stream...

Approving: None of my comments are blocking.

oulgen commented on torch/utils/_appending_byte_serializer.py:
```diff
@@ -0,0 +1,108 @@
+        self.write_int(len(payload))
+        self.write_bytes(payload)
+
+    def write_bytes(self, b: bytes) -> None:
```
if you look at the test cases, i have a write bytes example. I was thinking the serialize implementer would manually write the size for bytes since otherwise we run into cases where we have to mention size twice, i.e. the size of everything and just the next set of bytes. What do you think?

oulgen committed:
None on "Add AppendingByteSerializer class"


This PR adds a new util class that enables efficient appending of sequential byte data with custom serialization and deserialization. 

[ghstack-poisoned]

oulgen commented:
@pytorchbot merge
